github/messages.go

```diff
@@ -117,3 +147,44 @@ func validateSignature(signature string, payload, secretKey []byte) error {
 	}
 	return nil
 }
+
+// getWebhookEvent parses the payload and request, and build the Event.
+func getWebhookEvent(messageType string, payload []byte) (interface{}, error) {
+	eventType, ok := eventTypeMapping[messageType]
+	if !ok {
+		return nil, fmt.Errorf("unknown X-Github-Event in message: %v", messageType)
+	}
+
+	event := Event{
+		Type:       &eventType,
+		RawPayload: (*json.RawMessage)(&payload),
+	}
+	return event.Payload(), nil
+}
+
+// ParseWebhook parses the request, validates the signature, and returns
+// an interface to an activity event type (as returned by Event.Payload())
+//
+// secretKey is the GitHub Webhook secret message.
+//
+// Example usage:
+//
+//     func (s *GitHubEventMonitor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+//       event, err := github.ParseWebHook(r, s.webhookSecretKey)
+//       if err != nil { ... }
+//       switch event := event.(type) {
+//       case CommitCommentEvent:
+//           processCommitCommentEvent(event)
+//       case CreateEvent:
+//           processCreateEvent(event)
+//       ...
+//       }
+//     }
+//
+func ParseWebHook(r *http.Request, secretKey []byte) (interface{}, error) {
```

Comment:
I still have a concern and a suggestion about the API design here.

The entire implementation of `ParseWebHook` currently consists of calling an exported func `ValidatePayload`, then an unexported `getWebhookEvent`. Calling a public API can be a warning sign.

I'm worried that with the way this function is written now, it leads to poorly readable usage code. Consider seeing:

```Go
func (s *GitHubEventMonitor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	event, err := github.ParseWebHook(r, s.webhookSecretKey)
	if err != nil { ... }
```

If I'm looking over code, I will likely think that there may be a security issue because the webhook is being parsed without the payload being first validated. I'd have to look up the docs for and see that "ParseWebhook parses the request, validates the signature", so it's safe. But I think it would be much more readable if `ParseWebhook` had smaller scope and only did one thing:

```Go
// ParseWebhook takes the request header and a validated payload, and returns
// an interface to an activity event type (as returned by Event.Payload()).
func ParseWebHook(header http.Header, validatedPayload []byte) (interface{}, error) {
```

Then usage becomes very explicit and clearly safe/correct:

```Go
func (s *GitHubEventMonitor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	validatedPayload, err := github.ValidatePayload(r, s.webhookSecretKey)
	if err != nil { ... }
	event, err := github.ParseWebHook(r.Header, validatedPayload)
	if err != nil { ... }
```

Since this deals with security, I'd much prefer being explicit and having readable usage code. Verbosity is okay.

Alternatively, you could consider keeping the method functionality as is but rename it to `ValidateAndParseWebHook` to better reflect what it does.